modules/ocf_backups/files/upload-to-box

```diff
@@ -0,0 +1,444 @@
+#!/usr/bin/env python3
+import argparse
+import json
+import os
+import subprocess
+import sys
+import threading
+import time
+from collections import defaultdict
+from io import BytesIO
+from shutil import get_terminal_size
+
+import pycurl
+
+ARCHIVE = '/opt/backups/scratch/archive'
+BOX_DAV = 'https://dav.box.com/dav/'
+BOX_EMAILS = ['ckuehl@berkeley.edu', 'kpeng3.4@berkeley.edu']
+FOLDER = 'ocf-backup-' + time.strftime('%Y-%m-%d')
+MAX_CONCURR_UPLOADS = 10
+
+# To get the OAuth keys for an account:
+#
+# Mostly taken from https://goo.gl/bXdrPE
+#
+# Secrets like client_id, client_secret, and user email/password are in the
+# puppetmaster's private store and on hal in /opt/share/backups/box-creds.json.
+#
+# First, go to
+# https://account.box.com/api/oauth2/authorize?response_type=code&client_id={your_api_key}&state=authenticated&redirect_uri=https://www.ocf.berkeley.edu/oauth
+# You will get redirected to a 404 page, but you can get an auth code from that
+# page's URL. This auth code doesn't last very long at all, so run the below
+# command quickly. Otherwise, just go back to the same page and do it again.
+#
+# Quickly run: curl https://app.box.com/api/oauth2/token -X POST -d \
+# 'grant_type=authorization_code&code={auth code from above}&client_id={client id}&client_secret={client secret}'
+# Get the refresh token from this command's return, it can be used for up to
+# 60 days or one use before it expires. Once used, it is replaced by another
+# token, so it is stored in a local file (/opt/share/backupsbox-refresh-token)
+# and overwritten each time a new one is returned (when the old one is used).
+# Access tokens only last for an hour before expiring, so they are only stored
+# in this program's memory and are generated using the refresh token.
+
+
+class BoxUpload:
+
+    def __init__(self, filename, curl_auth):
+        self.filename = filename
+        self.auth = curl_auth
+        self.percentage = 0.0
+        self.bytes_uploaded = 0
+
+    def start(self, conn):
+        """Set curl params for the file to upload"""
+        file_path = ARCHIVE + '/' + self.filename
+        conn.upload = self
+
+        # PROGRESSFUNCTION is deprecated in pycurl >= 7.32.0, so this
+        # should be changed to conn.XFERINFOFUNCTION when we upgrade to stretch
+        conn.setopt(conn.PROGRESSFUNCTION, self.progress)
+        conn.setopt(conn.NOPROGRESS, False)
+
+        conn.file_bytes = os.path.getsize(file_path)
+        conn.fp = open(file_path, 'rb')
+
+        conn.setopt(conn.INFILESIZE, conn.file_bytes)
+        conn.setopt(conn.READFUNCTION, conn.fp.read)
+        conn.setopt(conn.UPLOAD, True)
+        conn.setopt(conn.URL, '{}{}/{}'.format(BOX_DAV, FOLDER, self.filename))
+        conn.setopt(conn.USERPWD, self.auth)
+
+    def progress(self, download_t, download_d, upload_t, upload_d):
```

Comment:
maybe call `store_progress`? current name implies it might return the current progress